### PR TITLE
fix: fix bad top1 accuracy in cifar_brevitas_training use case

### PR DESCRIPTION
**Pull request by RomanBredehoft**
We observed a small (~1.5% with 6 bits rounding) drop in accuracy when using a better (more representative) compilation inputset, which is currently being investigated : https://github.com/zama-ai/concrete-ml-internal/issues/3957

In the mean time, this should solve the very low top1 accuracy we currently face (17% -> 86% top1 accuracy with 6 bits rounding)  

closes https://github.com/zama-ai/concrete-ml-internal/issues/3939

### CODE REVIEW

File: use_case_examples/cifar_brevitas_training/evaluate_torch_cml.py [hunk 1]
```diff
@@ -97,9 +87,15 @@ def main(args):
     # Load weights
     model.load_state_dict(checkpoint["state_dict"], strict=False)
 
-    # Create a representative input-set that will be used used for both computing quantization
-    # parameters and compiling the model
-    input_set = torch.randn(1, 3, 32, 32)
```
Comment (RomanBredehoft) [inline use_case_examples/cifar_brevitas_training/evaluate_torch_cml.py:102]: this is the main change basically

File: use_case_examples/cifar_brevitas_training/bnn_pynq_train.py [hunk 2]
```diff
@@ -71,7 +71,7 @@ def parse_args(args):
     )
     add_bool_arg(parser, "detect_nan", default=False)
     # Compute resources
-    parser.add_argument("--num_workers", default=4, type=int, help="Number of workers")
+    parser.add_argument("--num_workers", default=0, type=int, help="Number of workers")
```
Comment (fd0r) [inline use_case_examples/cifar_brevitas_training/bnn_pynq_train.py:74]: 🙏🏼

**Review by fd0r — APPROVED** [review]

File: use_case_examples/cifar_brevitas_training/evaluate_torch_cml.py [hunk 3]
```diff
@@ -97,9 +87,15 @@ def main(args):
...
+    # Load the training set
+    train_set = get_train_set(dataset="CIFAR10", datadir=".datasets/")
+
+    # Create a representative input-set from the training set that will be used used for both
+    # computing quantization parameters and compiling the model
+    num_samples = 100
+    input_set = torch.stack(
+        [train_set[index][0] for index in range(min(num_samples, len(train_set)))]
+    )
```
Comment (jfrery) [inline use_case_examples/cifar_brevitas_training/evaluate_torch_cml.py:98]: Is the dataset seeded? Let's make sure we have the same calibration set so that we don't include randomness in the calibration.

File: use_case_examples/cifar_brevitas_training/bnn_pynq_train.py (hunk 2 above)
Comment (jfrery) [reply use_case_examples/cifar_brevitas_training/bnn_pynq_train.py:74]: Have you tried this on a linux machine? I remember this failed on mine but it's probably fixed now? I can try if you want.

**Review by jfrery — COMMENTED** [review]
Thanks! looks great to me.

File: use_case_examples/cifar_brevitas_training/bnn_pynq_train.py (hunk 2 above)
Comment (RomanBredehoft) [reply use_case_examples/cifar_brevitas_training/bnn_pynq_train.py:74]: yes if you could try that would be great, it fails on Mac if it's not set to 0

File: use_case_examples/cifar_brevitas_training/evaluate_torch_cml.py (hunk 3 above)
Comment (RomanBredehoft) [reply use_case_examples/cifar_brevitas_training/evaluate_torch_cml.py:98]: it has never been seeded in any use case if I recall, not sure we want to start to do that, especially since this inputset is probably less "random" than using the initial `torch.randn(1, 3, 32, 32)` 😅

File: use_case_examples/cifar_brevitas_training/trainer.py [hunk 4]
```diff
@@ -90,6 +91,61 @@ def accuracy(
         return list_topk_accs
 
 
+def get_train_set(dataset: str, datadir: Union[str, Path]) -> Dataset:
```
Comment (kcelia) [inline use_case_examples/cifar_brevitas_training/trainer.py:94]: It doesn't concern this PR, but we have a lot of redundancy between cifar_training and cifar_fietuning. Such as `get_train/test_set` in equivalent to `get_torchvision_dataset` in cifar_finetuning.

We should discuss the refactoring in the next meeting.

File: use_case_examples/cifar_brevitas_training/bnn_pynq_train.py (hunk 2 above)
Comment (jfrery) [reply use_case_examples/cifar_brevitas_training/bnn_pynq_train.py:74]: Looks like it's working for me

File: use_case_examples/cifar_brevitas_training/evaluate_torch_cml.py (hunk 3 above)
Comment (jfrery) [reply use_case_examples/cifar_brevitas_training/evaluate_torch_cml.py:98]: Yes that torch.randn was not great but the calibration data point was not having any impact on accuracy. Now it does so maybe we want to stick with one accuracy to be able to see any other causes of drop (else than random).

Comment (RomanBredehoft) [reply use_case_examples/cifar_brevitas_training/evaluate_torch_cml.py:98]: I don't think it matters a lot, the accuracy is stable with a 100 samples normalized between -1 and 1. Besides, we already have an issue for seeding notebooks : https://github.com/zama-ai/concrete-ml-internal/issues/1713

**Review by kcelia — APPROVED** [review]
